{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "70526a62",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "from icecream import ic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "a89513d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ic.disable()\n",
    "# '''Want to read sequences and display out of memory, but can keep stats in memory'''\n",
    "# def read_sequences(filename = 'genes.fa'):\n",
    "#     with open(filename, 'r') as file:\n",
    "#         # Read until line isn't read\n",
    "#         while (line := file.readline().strip()):\n",
    "#             ic(line)\n",
    "#             if line.startswith('>'): # Remake this condition part to a regex\n",
    "#                 title = line[1:]\n",
    "\n",
    "#             else:\n",
    "#                 display_seq_stats(title, *get_sequence_stats(line))\n",
    "                \n",
    "# read_sequences()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "5d82c0d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def splitter(fn):\n",
    "    def decorator(seq):\n",
    "        if not isinstance(seq, str):\n",
    "            return fn(seq)\n",
    "        \n",
    "        else:\n",
    "            re_split = re.split(r'([a-z]+)', seq)\n",
    "            return fn(re_split)\n",
    "    \n",
    "    return decorator\n",
    "\n",
    "@splitter\n",
    "def get_sequence_stats(re_split):\n",
    "    # Get statistics\n",
    "    exon_len = 0\n",
    "    num_exons = 0\n",
    "    intron_len = 0\n",
    "    num_introns = 0\n",
    "    for i in range(len(re_split)): \n",
    "        if (i+1) % 2 != 0: # Could check by sequence uppercase/lowercase preferably\n",
    "            exon_len += len(re_split[i])\n",
    "            num_exons += 1\n",
    "        else:\n",
    "            intron_len += len(re_split[i])\n",
    "            num_introns += 1\n",
    "    \n",
    "    total_len = exon_len + intron_len\n",
    "    return total_len, exon_len, intron_len, num_exons, num_introns\n",
    "\n",
    "\n",
    "def display_seq_stats(seq_name, total_len, exon_len, intron_len, num_exons, num_introns):\n",
    "    for elem in list(locals().values()):\n",
    "        print(f' {elem:^10} | ', end = ' ')\n",
    "    print()\n",
    "    \n",
    "def gene_stats_header():\n",
    "    gene = 'gene'\n",
    "    size = 'size'\n",
    "    exons_total_size = 'exons_len'\n",
    "    introns_total_size = 'intr_len'\n",
    "    exons_count = 'num_exons'\n",
    "    introns_count = 'num_intrs'\n",
    "    \n",
    "    display_seq_stats(*locals().values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "fd1e91eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_sequences(filename = 'genes.fa', gene_stats = False):\n",
    "    INTRON = 'Intron'\n",
    "    EXON = 'Exon'\n",
    "    with open(filename, 'r') as file:\n",
    "        # Read until line isn't read\n",
    "        if gene_stats:\n",
    "            stats = dict()\n",
    "        \n",
    "        while (line := file.readline().strip()):\n",
    "            exon_count = 1\n",
    "            intron_count = 1\n",
    "            if line.startswith('>'): # Remake this condition part to a regex\n",
    "                title = line[1:]\n",
    "                print(title, '\\n', '-'*20)\n",
    "            else:\n",
    "                re_split = re.split(r'([a-z]+)', line)\n",
    "                for i in range(len(re_split)):\n",
    "                    if (i+1) %2 != 0:\n",
    "                        print(f'{EXON:>10} {exon_count}: {re_split[i]}')\n",
    "                        exon_count += 1\n",
    "                    else:\n",
    "                        print(f'{INTRON:>10} {intron_count}: {re_split[i]}')\n",
    "                        intron_count += 1\n",
    "                stats[title] = get_sequence_stats(re_split)\n",
    "\n",
    "    return stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "b091807b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def main():\n",
    "    # Print gene intron/exon sequences\n",
    "    _ = '#' * 20\n",
    "    print(_, 'INTRON/EXON SEQUENCES PER GENE',_)\n",
    "    gene_stats = read_sequences(gene_stats = True) # prints sequences and returns stats\n",
    "    print('\\n'*3)\n",
    "    \n",
    "    # Display stats\n",
    "    gene_stats_header()\n",
    "    for k, v in gene_stats.items():\n",
    "        display_seq_stats(k, *v)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "5c457f33",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#################### INTRON/EXON SEQUENCES PER GENE ####################\n",
      "gene1 \n",
      " --------------------\n",
      "      Exon 1: TAGTGGTCTTTT\n",
      "    Intron 1: gagtg\n",
      "      Exon 2: TAGATCTGAAGGGAAAGTATTTCCACCAGTTCGGGGTCACCCAGCAGGGCAGGGTGACTTAAT\n",
      "gene2 \n",
      " --------------------\n",
      "      Exon 1: CGCGACTCGGCGCTCACAGTTAT\n",
      "    Intron 1: agcacgttt\n",
      "      Exon 2: AGACC\n",
      "    Intron 2: aaaaaa\n",
      "      Exon 3: CGGAGTTGGAT\n",
      "    Intron 3: cccgtgtg\n",
      "      Exon 4: AATCGGAGTCCTT\n",
      "gene3 \n",
      " --------------------\n",
      "      Exon 1: GTTACTTGTGAGCCTGGTT\n",
      "    Intron 1: agaggacgggagtggta\n",
      "      Exon 2: AATATAATTGTTGGC\n",
      "gene4 \n",
      " --------------------\n",
      "      Exon 1: AACATC\n",
      "    Intron 1: ttggttaa\n",
      "      Exon 2: AGGCTTTGATTAAACAATTTAAGCACGTAAATCCGAATTGACCTGATGACAATACGGAACATGCCGGCTCCGGG\n",
      "gene5 \n",
      " --------------------\n",
      "      Exon 1: ACCACCGGATAGGC\n",
      "    Intron 1: ttt\n",
      "      Exon 2: TGCT\n",
      "    Intron 2: aatgc\n",
      "      Exon 3: TATTAGGTCCAAAAGGTAGTATCGTAATAATGGCTCAGCCATGTCAATGTGCGGCATTCCAC\n",
      "gene6 \n",
      " --------------------\n",
      "      Exon 1: TAGATTCGAATCGATCGTGTTTCTCCCTCTGTGGGTTAACGAGGGGTCCG\n",
      "    Intron 1: cccccccgggggggggggggg\n",
      "      Exon 2: CTCGCATGTGCCGAACTTGTACCC\n",
      "gene7 \n",
      " --------------------\n",
      "      Exon 1: GAAATGGTTCGGTGCGATATCA\n",
      "    Intron 1: aa\n",
      "      Exon 2: GGCCGTT\n",
      "    Intron 2: ttt\n",
      "      Exon 3: CTCTTAACTTG\n",
      "    Intron 3: cccc\n",
      "      Exon 4: GCGGTGC\n",
      "    Intron 4: ggggg\n",
      "      Exon 5: AGATCC\n",
      "    Intron 5: aattg\n",
      "      Exon 6: GAACGTCTCTGGAGGGGTCGTGCGCTA\n",
      "gene8 \n",
      " --------------------\n",
      "      Exon 1: ATGTATACTAGACATTCTAACGCTCGCTTATTGGCGGAGACCATTTGCTCCACTACAAGAGGCTACTGT\n",
      "    Intron 1: aattgg\n",
      "      Exon 2: GTAGATCCGTA\n",
      "gene9 \n",
      " --------------------\n",
      "      Exon 1: TTCTTACACCCTTCTTTAGATCCAAA\n",
      "    Intron 1: aaaaaaaaaaaaaaaaaaaaaaaaaaatttttttttttttttgggggggggggggggggccccccccccccc\n",
      "      Exon 2: GTACCTCCATTTGCTCTGATGAC\n",
      "gene10 \n",
      " --------------------\n",
      "      Exon 1: CTACCT\n",
      "    Intron 1: atggctgat\n",
      "      Exon 2: CAACATCTACTAACGT\n",
      "    Intron 2: atggctgat\n",
      "      Exon 3: TCTGCCCTAACCTACAGGTCGATCCGAAATTCG\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "    gene    |      size    |   exons_len  |    intr_len  |   num_exons  |   num_intrs  |  \n",
      "   gene1    |       80     |       75     |       5      |       2      |       1      |  \n",
      "   gene2    |       75     |       52     |       23     |       4      |       3      |  \n",
      "   gene3    |       51     |       34     |       17     |       2      |       1      |  \n",
      "   gene4    |       88     |       80     |       8      |       2      |       1      |  \n",
      "   gene5    |       88     |       80     |       8      |       3      |       2      |  \n",
      "   gene6    |       95     |       74     |       21     |       2      |       1      |  \n",
      "   gene7    |       99     |       80     |       19     |       6      |       5      |  \n",
      "   gene8    |       86     |       80     |       6      |       2      |       1      |  \n",
      "   gene9    |      121     |       49     |       72     |       2      |       1      |  \n",
      "   gene10   |       73     |       55     |       18     |       3      |       2      |  \n"
     ]
    }
   ],
   "source": [
    "main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04b049f2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec4a7c99",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
